apps/ensindexer/src/handlers/Registrar.ts

```diff
@@ -81,6 +82,42 @@ export const makeRegistrarHandlers = ({
 
       const node = makeSubdomainNode(labelHash, registrarManagedNode);
 
+      // NOTE(preminted-names): The mainnet ENS Registrar(s) _always_ register a node with the ENS
+      // registry (emitting Registry#NewOwner) before emitting Registrar#NameRegistered.
+      //
+      // As part of the subgraph semantics, the Domain entity is created within Registry#NewOwner,
+      // and that Domain entity is expected to exist during the execution of these subsequent Registrar
+      // event handlers. This invariant is valid for the mainnet ENS contracts, which always register
+      // names with the Registry and emit the events as expected.
+      //
+      // This becomes an issue only because some plugins (Basenames, Lineanames) support the concept
+      // of 'preminting' a domain: these 'preminted' domains are reserved in the Registrar (which
+      // emits Registrar#NameRegistered) but a subname is NOT created in the Registry (which therefore
+      // does not emit Registry#NewOwner).
+      //
+      // https://github.com/base/basenames/blob/d00f71d822394cfaeab5aa7aded8225ef1292acc/src/L2/BaseRegistrar.sol#L248
+      // https://github.com/base/basenames/blob/d00f71d822394cfaeab5aa7aded8225ef1292acc/script/premint/Premint.s.sol#L36
+      //
+      // Preminted names are not 'real' ENS names — they do not exist in the Registry, and so we do
+      // not create a Domain entity for them. We also choose not to track their Registration, as that
+      // could result in unexpected behavior, when a Registration exists but the associated Domain
+      // does not.
+      //
+      // Therefore, if a Domain does not exist in these Registrar event handlers, it _must_ be a
+      // 'preminted' name, tracked only in the Registrar, and we ignore it. If/when these 'preminted'
+      // names are actually registered in the future, they will emit NewOwner as expected and the
+      // Domain and Registration entities will exist as normal.
+      const domain = await context.db.find(schema.domain, { id: node });
+      if (!domain) {
+        // no-op the NameRegistered event in plugins that explicitly support preminted names
+        if (pluginSupportsPremintedNames(pluginName)) return;
+
+        // invariant: if the domain does not exist and the plugin does not support preminted names, panic
+        throw new Error(
+          `Invariant: Registrar#NameRegistered was emitted before Registry#NewOwner and a Domain entity does not exist for node ${node}. This indicates that a name was registered in the Registrar but _not_ in the ENS Registry (i.e. 'preminted'). Currently this is only supported on Basenames and Lineanames, but this occurred in plugin '${pluginName}'.`,
+        );
+      }
```

Comment:
Nice code branch reversal to execute guard expressions before attempting to update a possibly not existing domain entity 👍